admin/compressor/class-boldgrid-backup-admin-compressor-system-zip-test.php

```diff
@@ -0,0 +1,300 @@
+			);
+
+			return false;
+		}
+
+		// Create all the files. Abort if one could not be.
+		foreach ( $this->items as $item ) {
+			$created = false;
+
+			switch ( $item['type'] ) {
+				case 'd':
+					$created = $this->core->wp_filesystem->mkdir( $this->test_dir . DIRECTORY_SEPARATOR . $item['name'] );
+					break;
+				case 'f':
+					$created = $this->core->wp_filesystem->touch( $this->test_dir . DIRECTORY_SEPARATOR . $item['name'] );
+					break;
+			}
+
+			// Add this item to our filelist (a .txt file containing a list of files to backup).
+			$this->filelist[] = $item['name'];
+
+			// Above if we could not create a file.
+			if ( ! $created ) {
+				$this->error = sprintf(
+					// translators: 1 The name of a file that could not be created.
+					esc_html__( 'Unable to create test file: %1$s', 'boldgrid-backup' ),
+					$item['name']
+				);
+
+				return false;
+			}
+		}
+
+		// Save the filelist.
+		if ( ! $this->core->wp_filesystem->put_contents( $this->filelist_filepath, implode( PHP_EOL, $this->filelist ) ) ) {
+			$this->error = esc_html__( 'Unable to create filelist.txt', 'boldgrid-backup' );
+
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Restore our test directory from zip and make sure everything was restored.
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @return bool True on success.
+	 */
+	private function test_dir_restore() {
+		// Delete the folder we zipped and make sure it's gone.
+		if ( ! $this->core->wp_filesystem->delete( $this->test_dir, true ) ) {
+			$this->error = esc_html__( 'Unable to delete test directory.', 'boldgrid-backup' );
+			return false;
+		}
+
+		// Create our test directory again.
+		if ( ! $this->core->wp_filesystem->mkdir( $this->test_dir ) ) {
+			$this->error = esc_html__( 'Unable to recreate test directory.', 'boldgrid-backup' );
+			return false;
+		}
+
+		// Move our zip file to our test directory.
+		if ( ! $this->core->wp_filesystem->move( $this->zip_filepath, $this->test_dir . DIRECTORY_SEPARATOR . basename( $this->zip_filepath ) ) ) {
+			$this->error = esc_html__( 'Unable to move zip to test directory.', 'boldgrid-backup' );
+			return false;
+		}
+
+		// Unzip.
+		$this->core->execute_command( 'cd ' . $this->test_dir . '; unzip ' . basename( $this->zip_filepath ) );
+		foreach ( $this->items as $item ) {
+			$path = $this->test_dir . DIRECTORY_SEPARATOR . $item['name'];
+
+			if ( ! $this->core->wp_filesystem->exists( $path ) ) {
+				$this->error = esc_html__( 'Unable to restore all files and folders.', 'boldgrid-backup' );
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Zip up our test directory.
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @return bool True on success.
+	 */
+	private function test_dir_zip() {
+		// Zip everything up and make sure it worked.
+		$this->core->execute_command( 'cd ' . $this->test_dir . '; zip ' . $this->zip_filepath . ' -@ < ' . $this->filelist_filepath );
+		if ( ! $this->core->wp_filesystem->exists( $this->zip_filepath ) ) {
```

Comment:
When this test is run on a server without the zip binary installed, the only error returned is 'Unable to create zip file' with no specifier as to why. Adding a test for whether or not the zip command exists would help with people troubleshooting why this option doesn't work. 

Additionally, when this test failed for me, the Preflight Check status was listed as FAILED. This failure shouldn't cause the whole functionality test to fail unless the PclZip and ZipArchive both failed.

Reply:
>Additionally, when this test failed for me, the Preflight Check status was listed as FAILED. This failure shouldn't cause the whole functionality test to fail unless the PclZip and ZipArchive both failed.

Resolved by https://github.com/BoldGrid/boldgrid-backup/pull/205/commits/ba46dae11511befb24ccceabd304a21447cd50f1 and https://github.com/BoldGrid/boldgrid-backup/pull/205/commits/4f67c6d3eb2788fbb9c93f77fe2f430fabc88bde

Reply:
>When this test is run on a server without the zip binary installed, the only error returned is 'Unable to create zip file' with no specifier as to why. Adding a test for whether or not the zip command exists would help with people troubleshooting why this option doesn't work.

Resolved by https://github.com/BoldGrid/boldgrid-backup/commit/4f67c6d3eb2788fbb9c93f77fe2f430fabc88bde and https://github.com/BoldGrid/boldgrid-backup/pull/205/commits/bae0c1b1935431c68f511e9407466bd4b1ab2ad5